File: frontend/src/types/wall-textures.ts
/**
 * Wall Texture types for the design studio
 *
 * Used for the Asian Paints (and other vendors) textured wall finishes
 * and visualization feature. Textures are grouped by base name with
 * multiple color variants per texture.
 */

import { WallColorFamily } from './wall-colors';

/**
 * Texture type enum for categorizing wall textures
 */
export type TextureType =
  | 'marble'
  | 'velvet'
  | 'stone'
  | 'concrete'
  | '3d'
  | 'wall_tile'
  | 'stucco'
  | 'rust'
  | 'other';

/**
 * Human-readable labels for texture types
 */
export const TEXTURE_TYPE_LABELS: Record<TextureType, string> = {
  marble: 'Marble',
  velvet: 'Velvet',
  stone: 'Stone',
  concrete: 'Concrete',
  '3d': '3D',
  wall_tile: 'Wall Tile',
  stucco: 'Stucco',
  rust: 'Rust',
  other: 'Other',
};

/**
 * Order for displaying texture types
 */
export const TEXTURE_TYPE_ORDER: TextureType[] = [
  'marble',
  'velvet',
  'stone',
  'concrete',
  '3d',
  'wall_tile',
  'stucco',
  'rust',
  'other',
];

/**
 * Individual texture variant
 * Each variant is a separate product with its own swatch image and product page.
 * image_data is optional — the list endpoint returns variants WITHOUT inline images
 * to keep the response small. Use the image endpoint URL for <img> tags instead.
 */
export interface WallTextureVariant {
  id: number;
  code: string;                    // Variant code (e.g., "TNB1003CMB1001")
  name?: string;                   // Variant name if different from parent
  image_data?: string;             // Base64 encoded texture swatch image (only from individual variant endpoint)
  image_url?: string;              // Original source URL for swatch image
  product_url?: string;            // Product page URL for this variant
  color_family?: WallColorFamily;  // For color filtering
  is_active: boolean;
  display_order: number;
}

/**
 * Base wall texture (without variants)
 */
export interface WallTexture {
  id: number;
  name: string;                    // Texture name (e.g., "Basket", "Bandhej")
  collection?: string;             // Collection name (e.g., "Lux Imprints")
  texture_type?: TextureType;      // Type of texture finish
  brand: string;                   // Brand/vendor (e.g., "Asian Paints")
  description?: string;
  is_active: boolean;
  display_order: number;
}

/**
 * Wall texture with all its color variants
 */
export interface WallTextureWithVariants extends WallTexture {
  variants: WallTextureVariant[];
}

/**
 * Texture type metadata for filter UI
 */
export interface TextureTypeInfo {
  value: TextureType;
  label: string;
  texture_count: number;
}

/**
 * Texture brand metadata for filter UI
 */
export interface TextureBrandInfo {
  name: string;
  texture_count: number;
}

/**
 * Full textures response from API with filter metadata
 */
export interface WallTexturesGroupedResponse {
  textures: WallTextureWithVariants[];
  brands: TextureBrandInfo[];
  texture_types: TextureTypeInfo[];
  total_count: number;
}

/**
 * Request to change wall texture in visualization
 */
export interface ChangeWallTextureRequest {
  room_image: string;           // Base64 encoded current visualization image
  texture_variant_id: number;   // ID of the texture variant to apply
  user_id?: string;
  session_id?: string;
}

/**
 * Response from wall texture change API
 */
export interface ChangeWallTextureResponse {
  success: boolean;
  rendered_image?: string;      // Base64 encoded result image
  error_message?: string;
  processing_time: number;
  texture_name?: string;
  texture_type?: string;
}

/**
 * Wall type toggle options (for Panel 1 filter)
 */
export type WallType = 'color' | 'textured';

/**
 * Wall filter state for managing filter selections
 */
export interface WallFilterState {
  wallType: WallType;
  // Color filters
  selectedFamilies: WallColorFamily[];
  // Texture filters
  selectedBrands: string[];
}

/**
 * Default wall filter state
 */
export const DEFAULT_WALL_FILTER_STATE: WallFilterState = {
  wallType: 'color',
  selectedFamilies: [],
  selectedBrands: [],
};

File: api/schemas/wall_textures.py
"""
Pydantic schemas for wall texture API endpoints

Provides schemas for textures, texture variants, and visualization requests.
Textures are grouped by base name with multiple color variants per texture.
"""
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field
from schemas.wall_colors import WallColorFamily


class TextureType(str, Enum):
    """Texture finish types for wall textures"""

    MARBLE = "marble"
    VELVET = "velvet"
    STONE = "stone"
    CONCRETE = "concrete"
    THREE_D = "3d"
    WALL_TILE = "wall_tile"
    STUCCO = "stucco"
    RUST = "rust"
    OTHER = "other"


# Human-readable labels for texture types
TEXTURE_TYPE_LABELS = {
    TextureType.MARBLE: "Marble",
    TextureType.VELVET: "Velvet",
    TextureType.STONE: "Stone",
    TextureType.CONCRETE: "Concrete",
    TextureType.THREE_D: "3D",
    TextureType.WALL_TILE: "Wall Tile",
    TextureType.STUCCO: "Stucco",
    TextureType.RUST: "Rust",
    TextureType.OTHER: "Other",
}


class WallTextureVariantSchema(BaseModel):
    """Individual texture variant with image data"""

    id: int
    code: str
    name: Optional[str] = None
    image_data: str = Field(..., description="Base64 encoded room/wall shot image")
    image_url: Optional[str] = None
    swatch_url: Optional[str] = Field(None, description="URL for the texture swatch pattern image")
    color_family: Optional[WallColorFamily] = None
    is_active: bool = True
    display_order: int = 0

    class Config:
        from_attributes = True


class WallTextureVariantLightSchema(BaseModel):
    """Lightweight texture variant without inline image data (for list endpoints)"""

    id: int
    code: str
    name: Optional[str] = None
    image_url: Optional[str] = None
    swatch_url: Optional[str] = None
    color_family: Optional[WallColorFamily] = None
    is_active: bool = True
    display_order: int = 0

    class Config:
        from_attributes = True


class WallTextureSchema(BaseModel):
    """Base wall texture without variants"""

    id: int
    name: str
    collection: Optional[str] = None
    texture_type: Optional[TextureType] = None
    brand: str = "Asian Paints"
    description: Optional[str] = None
    is_active: bool = True
    display_order: int = 0

    class Config:
        from_attributes = True


class WallTextureWithVariantsSchema(BaseModel):
    """Wall texture with all its color variants (includes image data)"""

    id: int
    name: str
    collection: Optional[str] = None
    texture_type: Optional[TextureType] = None
    brand: str = "Asian Paints"
    description: Optional[str] = None
    variants: List[WallTextureVariantSchema] = []

    class Config:
        from_attributes = True


class WallTextureWithVariantsLightSchema(BaseModel):
    """Wall texture with lightweight variants (no inline images - for list endpoints)"""

    id: int
    name: str
    collection: Optional[str] = None
    texture_type: Optional[TextureType] = None
    brand: str = "Asian Paints"
    description: Optional[str] = None
    variants: List[WallTextureVariantLightSchema] = []

    class Config:
        from_attributes = True


class TextureTypeSchema(BaseModel):
    """Texture type with label"""

    value: TextureType
    label: str
    texture_count: int = 0


class TextureBrandSchema(BaseModel):
    """Texture brand with count"""

    name: str
    texture_count: int = 0


class WallTexturesGroupedResponse(BaseModel):
    """Wall textures with metadata for filters (lightweight, no inline images)"""

    textures: List[WallTextureWithVariantsLightSchema]
    brands: List[TextureBrandSchema]
    texture_types: List[TextureTypeSchema]
    total_count: int = 0


class ChangeWallTextureRequest(BaseModel):
    """Request to change wall texture in visualization"""

    room_image: str = Field(..., description="Base64 encoded current visualization image")
    texture_variant_id: int = Field(..., description="ID of the texture variant to apply")


class ChangeWallTextureResponse(BaseModel):
    """Response from wall texture change"""

    success: bool
    rendered_image: Optional[str] = Field(None, description="Base64 encoded result image")
    error_message: Optional[str] = None
    processing_time: float = Field(0.0, description="Processing time in seconds")
    texture_name: Optional[str] = None
    texture_type: Optional[str] = None
